src/data_structures/interval_tree.rs

```diff
@@ -0,0 +1,315 @@
+extern crate num;
+
+use self::num::traits::Num;
+
+use std::cmp;
+use std::mem;
+use std::fmt::Debug;
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct IntervalTree<N, D> {
+    root: Option<Node<N, D>>,
+}
+
+impl<N: Debug + Num + Clone + Ord, D: Debug> IntervalTree<N, D> {
+    pub fn new() -> Self {
+        IntervalTree { root: None }
+    }
+
+    pub fn insert(&mut self, start: N, end: N, data: D) {
+        validate_interval(&start, &end);
+        match self.root {
+            Some(ref mut n) => n.insert(start, end, data),
+            None => self.root = Some(Node::new(start, end, data)),
+        }
+    }
+
+    pub fn find_intersection(&self, start: N, end: N) -> Option<&D> {
```

Comment:
Yes, reference sounds reasonable. When there are multiple intervals, an arbitrary one would be returned. I think returning all intersections is a good idea, I was also thinking about this but didn't code that yet. As of returning Node, there's much tree metadata there user shouldn't really care about, so maybe something like a (start: N, end: N, data: D) tuple?